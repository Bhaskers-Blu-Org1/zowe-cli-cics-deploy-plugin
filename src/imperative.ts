/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
* Copyright IBM Corp, 2019
*
*/

import {IImperativeConfig} from "@brightside/imperative";

const MAX_LENGTH = 8;
const MAX_HLQ_LENGTH = 35;

const config: IImperativeConfig = {
    commandModuleGlobs: ["**/cli/*/*.definition!(.d).*s"],
    pluginHealthCheck: __dirname + "/healthCheck.handler",
    pluginSummary: "Generate and deploy IBM CICS bundle resources",
    pluginAliases: ["cdep"],
    rootCommandDescription: "CICS bundle deployment plugin.",
    productDisplayName: "Zowe cics-deploy plug-in",
    name: "cics-deploy",
    profiles: [
      {
        type: "cics-deploy",
        schema: {
          type: "object",
          title: "The cics-deploy-profile schema",
          description: "Specifies the target environment for the cics-deploy deploy and undeploy actions.",
          properties: {
            cicsplex: {
              optionDefinition: {
                description: "Specifies the CICSplex (up to 8 characters) to target.",
                type: "string",
                name: "cicsplex",
                aliases: ["cp"],
                stringLengthRange: [1, MAX_LENGTH],
                required: true
              },
              type: "string"
            },
            scope: {
              optionDefinition: {
                description: "Specifies the name of the CICS System, or CICS System Group " +
                             "(up to 8 characters) to target.",
                type: "string",
                name: "scope",
                aliases: ["sc"],
                stringLengthRange: [1, MAX_LENGTH],
                impliesOneOf: ["csdgroup", "resgroup"],
                required: true
              },
              type: "string"
            },
            csdgroup: {
              optionDefinition: {
                description: "Specifies the CSD group (up to 8 characters) for the bundle resource. If a bundle is " +
                             "deployed then a definition is added to this group; if a bundle is undeployed then the " +
                             "definition is removed from this group. The CSD group is changed for each CICS system " +
                             "that is specified by the --scope option. The --csdgroup and --resgroup options are " +
                             "mutually exclusive.",
                type: "string",
                name: "csdgroup",
                aliases: ["cg"],
                stringLengthRange: [1, MAX_LENGTH]
              },
              type: "string"
            },
            resgroup: {
              optionDefinition: {
                description: "Specifies the BAS resource group (up to 8 characters) for the bundle resource. If a bundle is " +
                             "deployed then a resource is defined in the BAS data repository; if a bundle is undeployed then the " +
                             "definition is removed. The --csdgroup and --resgroup options are mutually exclusive.",
                type: "string",
                name: "resgroup",
                aliases: ["rg"],
                stringLengthRange: [1, MAX_LENGTH],
                conflictsWith: [ "csdgroup" ],
              },
              type: "string"
            },
            cicshlq: {
              optionDefinition: {
                description: "Specifies the High Level Qualifier (up to 35 characters) at which the CICS " +
                             "datasets can be found in the target environment.",
                type: "string",
                name: "cicshlq",
                aliases: ["hlq"],
                stringLengthRange: [1, MAX_HLQ_LENGTH],
                required: true
              },
              type: "string"
            },
            cpsmhlq: {
              optionDefinition: {
                description: "Specifies the High Level Qualifier (up to 35 characters) at which the CPSM " +
                             "datasets can be found in the target environment.",
                type: "string",
                name: "cpsmhlq",
                aliases: ["cphlq"],
                stringLengthRange: [1, MAX_HLQ_LENGTH],
                required: true
              },
              type: "string"
            },
            jobcard: {
              optionDefinition: {
                description: "Specifies the job card to use with any generated DFHDPLOY JCL.",
                type: "string",
                name: "jobcard",
                defaultValue: "//DFHDPLOY JOB DFHDPLOY,CLASS=A,MSGCLASS=X,TIME=NOLIMIT",
                aliases: ["jc"],
                required: true
              },
              type: "string"
            }
          },
          required: ["cicsplex", "scope", "cicshlq", "cpsmhlq", "jobcard"]
        }
      }
    ]
};

export = config;
