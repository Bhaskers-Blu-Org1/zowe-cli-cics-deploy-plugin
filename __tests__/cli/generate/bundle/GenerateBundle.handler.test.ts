/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
* Copyright IBM Corp, 2019
*
*/

import {CheckStatus, ZosmfSession} from "@zowe/cli";
import {IHandlerParameters, Imperative, ImperativeError} from "@zowe/imperative";
import * as GenerateBundleDefinition from "../../../../src/cli/generate/bundle/GenerateBundle.definition";
import * as GenerateBundleHandler from "../../../../src/cli/generate/bundle/GenerateBundle.handler";
import * as fs from "fs";

process.env.FORCE_COLOR = "0";

const DEFAULT_PARAMTERS: IHandlerParameters = {
    arguments: {
        $0: "bright",
        _: ["zowe-cli-cics-deploy-plugin", "generate", "bundle"],
        silent: true
    },
    profiles: {
        get: (type: string) => {
            return {};
        }
    } as any,
    response: {
        data: {
            setMessage: jest.fn((setMsgArgs) => {
                expect("" + setMsgArgs).toMatch("NO RESPONSE MESSAGE IS EXPECTED");
            }),
            setObj: jest.fn((setObjArgs) => {
                expect(setObjArgs).toMatch("NO RESPONSE OBJECT IS EXPECTED");
            })
        },
        console: {
            log: jest.fn((logs) => {
                expect("" + logs).toContain("CICS Bundle");
                expect("" + logs).toContain("generated");
            }),
            error: jest.fn((errors) => {
                expect("" + errors).toMatch("NO ERROR MESSAGE IS EXPECTED");
            }),
            errorHeader: jest.fn(() => undefined)
        },
        progress: {
            startBar: jest.fn((parms) => undefined),
            endBar: jest.fn(() => undefined)
        }
    } as any,
    definition: GenerateBundleDefinition.GenerateBundleDefinition,
    fullDefinition: GenerateBundleDefinition.GenerateBundleDefinition,
};

describe("bundle Handler", () => {
    afterEach(() => {
        jest.restoreAllMocks();
    });
    it("should process the current directory", async () => {
        DEFAULT_PARAMTERS.arguments.nosave = "true";

        let error;
        try {
          const handler = new GenerateBundleHandler.default();
          // The handler should succeed
          const params = Object.assign({}, ...[DEFAULT_PARAMTERS]);
          await handler.process(params);
        } catch (e) {
            error = e;
            Imperative.console.error(`Error experienced: ${e.message}`);
        }
        expect(error).toBeUndefined();
    });
    it("should cope with an empty directory", async () => {
        DEFAULT_PARAMTERS.arguments.nosave = "true";

        const currentDir = process.cwd();
        if (!fs.existsSync("__tests__/__resources__/EmptyBundle01")) {
            fs.mkdirSync("__tests__/__resources__/EmptyBundle01");
        }
        process.chdir("__tests__/__resources__/EmptyBundle01");

        let error;
        try {
          const handler = new GenerateBundleHandler.default();
          // The handler should succeed
          const params = Object.assign({}, ...[DEFAULT_PARAMTERS]);
          await handler.process(params);
        } catch (e) {
            error = e;
            Imperative.console.error(`Error experienced: ${e.message}`);
        }
        process.chdir(currentDir);
        expect(error).toBeUndefined();
    });
    it("should process a directory with package.json", async () => {
        DEFAULT_PARAMTERS.arguments.nosave = "true";

        const currentDir = process.cwd();
        process.chdir("__tests__/__resources__/ExampleBundle04");

        let error;
        try {
          const handler = new GenerateBundleHandler.default();
          // The handler should succeed
          const params = Object.assign({}, ...[DEFAULT_PARAMTERS]);
          await handler.process(params);
        } catch (e) {
            error = e;
            Imperative.console.error(`Error experienced: ${e.message}`);
        }
        process.chdir(currentDir);
        expect(error).toBeUndefined();
    });
});
