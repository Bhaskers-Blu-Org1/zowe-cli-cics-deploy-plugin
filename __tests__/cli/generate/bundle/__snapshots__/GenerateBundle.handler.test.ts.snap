// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bundle Handler should process the current directory 1`] = `"NODEJSAPP \\"zowe-cli-cics-deploy-plugin\\" defined for startscript \\"lib/index.js\\""`;

exports[`bundle Handler should process the current directory 2`] = `"CICS Bundle \\"zowe-cli-cics-deploy-plugin\\" generated"`;
